File: Medium/Kernel-Adventures-Part-1/construct_pwd.py
"""Construct a password that hashes to a target value.
Using the property that hash_round is a bijection for each character."""

import sys

def inv_xor_shift_6(y):
    x = y & 0xfc000000
    for i in range(25, -1, -1):
        x |= (((y >> i) ^ (x >> (i + 6))) & 1) << i
    return x & 0xffffffff

INV_1025 = pow(1025, -1, 2**32)

def hash_round(r, c):
    c = c & 0xff
    r = (r + c) & 0xffffffff
    r = (r * 1025) & 0xffffffff
    r ^= (r >> 6)
    r ^= c
    return r

def hash_str(s):
    r = 0
    for ch in s:
        r = hash_round(r, ch)
    return r

def construct_password(target, length):
    """Build a password of given length that hashes to target."""
    # Go backwards from target for length-1 steps with arbitrary chars
    # Then find the final char that connects r=0 to the intermediate value
    
    r = target
    chars = []
    
    # length-1 arbitrary backwards steps
    for step in range(length - 1):
        c = (step % 94) + 33
        r_temp = r ^ c
        r2 = inv_xor_shift_6(r_temp)
        r1 = (r2 * INV_1025) & 0xffffffff
        r_in = (r1 - c) & 0xffffffff
        verify = hash_round(r_in, c)
        if verify != r:
            raise ValueError(f"Step {step}: verification failed")
        chars.append(c)
        r = r_in
    
    # Now r = r_1, the value after first forward character
    # Find c such that hash_round(0, c) = r_1
    last_char = None
    for c in range(1, 256):
        if hash_round(0, c) == r:
            last_char = c
            break
    
    if last_char is None:
        raise ValueError(f"Cannot find final char for r=0 -> 0x{r:08x}, try different length")
    
    # Password = [last_char] + reverse(chars)
    password = bytes([last_char] + list(reversed(chars)))
    
    # Verify
    r_test = 0
    for ch in password:
        r_test = hash_round(r_test, ch)
    assert r_test == target, f"Forward check: {r_test} != {target}"
    
    return password

if __name__ == '__main__':
    targets = [
        ("users[1]", 53623157),
        ("users[3]", 716789863),
    ]
    for name, target in targets:
        print(f"\n=== {name}: {target} (0x{target:08x}) ===")
        for length in [1, 2, 3, 5, 10, 16]:
            try:
                pwd = construct_password(target, length)
                v = hash_str(pwd)
                ok = "OK" if v == target else "FAIL"
                print(f"  len={length}: {pwd!r} verify={v} {ok}")
            except ValueError as e:
                print(f"  len={length}: {e}")

File: Medium/Kernel-Adventures-Part-1/construct_pwd.c
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>

uint32_t hash_round(uint32_t r, uint8_t c) {
    r += c;
    r *= 1025;
    r ^= (r >> 6);
    r ^= c;
    return r;
}

uint32_t inv_xor_shift_6(uint32_t y) {
    uint32_t x = y & 0xfc000000;
    for (int i = 25; i >= 0; i--) {
        uint32_t bit = ((y >> i) ^ (x >> (i + 6))) & 1;
        x |= (bit << i);
    }
    return x;
}

#define INV_1025 3222273025u

uint32_t inv_hash_round(uint32_t r_out, uint8_t c) {
    uint32_t r_temp = r_out ^ c;
    uint32_t x = inv_xor_shift_6(r_temp);
    uint32_t r1 = x * INV_1025;
    return r1;
}

#define TARGET 716661863u
#define PWDLEN 100  // 100-char password for wide race window

int main() {
    uint8_t pwd[PWDLEN + 1] = {0};
    
    // Step 1: go backwards from target for (PWDLEN-1) steps with random chars
    uint32_t r = TARGET;
    for (int i = 0; i < PWDLEN - 1; i++) {
        uint8_t c = (rand() % 255) + 1;
        pwd[PWDLEN - 1 - i] = c;
        r = inv_hash_round(r, c);
    }
    
    // Step 2: find final char that connects r to hash_round(0, c)
    for (int c = 1; c < 256; c++) {
        if (hash_round(0, c) == r) {
            pwd[0] = c;
            // Verify
            uint32_t r_test = 0;
            for (int i = 0; i < PWDLEN; i++)
                r_test = hash_round(r_test, pwd[i]);
            if (r_test == TARGET) {
                printf("Password length %d:\n", PWDLEN);
                for (int i = 0; i < PWDLEN; i++)
                    printf("\\x%02x", pwd[i]);
                printf("\n\nC array:\nstatic const uint8_t password[] = {");
                for (int i = 0; i < PWDLEN; i++) {
                    if (i > 0) printf(", ");
                    printf("0x%02x", pwd[i]);
                }
                printf(", 0};\n");
                return 0;
            }
        }
    }
    printf("Failed to find connecting char (very unlikely)\n");
    return 1;
}
